Lock Screen Widgets: Create Two Field Widget

staskus opened the pull request:
Fixes #21489

## Description

1. Created `LockScreenMultiStatView` to show two fields, reusing elements from `LockScreenSingleStatView`
2. Created `LockScreenMultiStatWidgetViewProvider` that configures all the states of the multi-stat widget, made providers more generic to support different types of `HomeWidgetData`
3. Created `LockScreenTodayLikesCommentsStatWidgetConfig`, to add `Likes and Comments` widget
4. Other changes are related to refactoring that cascades throughout multiple files

I plan to add more single and multi-field widgets in other PRs. For now, added one for Likes and Comments

## To test:

### Happy Path

1. Open Jetpack
2. Log into the account
3. Enable Lock Screen Widget feature flag
4. Open Lock screen, long tap, tap Customize, select Widget area
5. Tap on Jetpack
6. Confirm 2 "Today" Jetpack widgets exists, confirm one shows "Views", another "Likes and Comments"
7. Add Both
8. Tap on one of the widgets to change the site
9. Confirm that the data loads as expected
10. Confirm tapping on the widget opens "Today" Stats view

### Logged out

1. After Happy-Path, logout
2. Confirm widgets show "Please login" option

### No site

1. After Happy-Path, logout
2. Log into account with no sites
3. Confirm widgets show "No site" message

## Regression Notes
1. Potential unintended areas of impact

Breaking single stat widget

12. What I did to test those areas of impact (or what existing automated tests I relied on)

Manual testing, running unit-tests

13. What automated tests I added (or what prevented me from doing so)

None

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:
- [x] Portrait and landscape orientations.
- [x] Light and dark modes.
- [x] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [x] VoiceOver.
- [x] Languages with large words or with letters/accents not frequently used in English.
- [x] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [x] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

## Images

<img width="290" alt="long numbers 3" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/915480df-f49e-4636-89f9-b29cb93b844a">
<img width="292" alt="long numbers 2" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/ee64f732-2442-478e-878c-bb240dfca165">
<img width="362" alt="image" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/ee725434-da37-41b5-b59a-9e33fffd27fb">
<img width="506" alt="image" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/043aa4fe-d163-4333-a8d7-39c5b5620b21">
<img width="330" alt="image" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/0264ba5d-4226-41f8-b415-c0828b0d5568">

staskus committed:
Created LockScreenMultiStatView to display two field lock screen widget

staskus committed:
Created configuration for Today Views and Comments lock screen widget

staskus committed:
Use the same "Today" placeholder title for all Today lock screen widgets

staskus committed:
Change multi field element to show Likes and Comments

staskus committed:
Switch to a horizontal layout for multi stat view

staskus committed:
Pass value as Int for both single and multi field views

staskus committed:
Make title fonts smaller to fit longer text

staskus committed:
Refactor the code so ViewProviders would accept generic data

staskus committed:
WordPress test fixes

staskus commented:
@guarani 

You can see in this draft PR https://github.com/wordpress-mobile/WordPress-iOS/pull/21495 how more configurations are added, and how they would look. I provided the video.

With more configurations, I also found some issues with labels fitting when both numbers are in the hundreds of thousands, and titles do not always fit as well. I will look for more clever ways to ensure they fit since setting lower `minimumScaleFactor` could result in one field shrinking and another field staying at the same size which is undesirable.

guarani commented on WordPress/WordPressStatsWidgets/LockScreenWidgets/Models/LockScreenMultiStatViewModel.swift:
```diff
@@ -0,0 +1,13 @@
+
+    let siteName: String
+    let updatedTime: Date
+    let firstField: Field
```
nit: this could be named `primary` to match `secondary` below

guarani commented on WordPress/WordPressStatsWidgets/LockScreenWidgets/ViewProvider/LockScreenMultiStatWidgetViewProvider.swift:
```diff
@@ -0,0 +1,78 @@
+            case .thisWeek:
+                return AppConfiguration.Widget.Localization.unconfiguredViewThisWeekTitle
+            }
+        }
```
nit: a `let` could be used here, which results in less nesting:
```
        let message: String
        switch widgetKind {
        case .today:
            message = AppConfiguration.Widget.Localization.unconfiguredViewTodayTitle
        case .allTime:
            message = AppConfiguration.Widget.Localization.unconfiguredViewAllTimeTitle
        case .thisWeek:
            message = AppConfiguration.Widget.Localization.unconfiguredViewThisWeekTitle
        }
```

guarani commented on WordPress/WordPressStatsWidgets/Model/HomeWidgetAllTimeData.swift:
```diff
@@ -8,4 +8,9 @@ struct HomeWidgetAllTimeData: HomeWidgetData {
+
+    var statsURL: URL? {
+        let statsUrl = "https://wordpress.com/stats/insights/"
+        return URL(string: statsUrl + "\(siteID)")
```
nit: to safe-guard against the trailing slash being accidentally removed, the path could be appended like this:
```
let statsUrl = URL(string: "https://wordpress.com/stats/insights/")
return statsUrl.appending(path: String(siteID))
```

guarani commented on WordPress/WordPressStatsWidgets/Tracks/Tracks+StatsWidgets.swift:
nit: This line at the top of the file:

> /// This extension implements helper tracking methods, meant for Today Home Widget usage.

is outdated since it talks about Today widgets. It also could say that the file is for Lock screen widgets. It could be changed to:

/// This extension implements helper tracking methods meant for Home & Lock screen widgets.

guarani reviewed (commented):
I couldn't get the widgets to show up on my device and I'm not sure why. I enabled the feature flag. 

However, since I also tried a brand new test project with a sample lock screen widget and it didn't show up there, it might be an issue with my setup. I ran the Jetpack target but no luck today.

I'll come back to this tomorrow 👋

staskus committed:
Renaming firstField to primaryField

staskus committed:
Use let instead of var for message

staskus committed:
Appending path component more safely using URL method

staskus committed:
Update Tracks+StatsWidgets.swift

staskus committed:
Configure consistent font sizes for different types of value lengths

- Using allowsTightening flag to fit more text for the same font size
- Adding Spacers so widgets with different font sizes have elements aligned at the same positions
- Adding manual font size calculation for multi-field widget so both Text fields would have same font size when the value is longer

staskus commented:
Made additional improvements to make sure widget elements are aligned and consistent with different font sizes

https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/eb6aaa62-a7c0-44da-a48d-6248c32861c1

guarani reviewed (approved):
👋 I was able to get this working by removing the feature flag (see discussion in p1694005726199069-slack-C441E3YTS where we found that feature flags were not working in extensions).

#### Happy Path testing

This worked great 👍 I tested with a couple of sites.

#### Logged Out testing

I saw the message asking me to log in here.

#### No Sites testing

(There might be a typo because this is called "Logged out" testing above). Otherwise this worked and I saw the message asking me to create a site. 

---

I noticed a few scenarios where the widgets got "stuck" showing an outdated message. For example:
1. Log out of the app and notice the app shows the "Log in..." message as expected
2. Log in again and the Views widget shows data but the Likes & Comments one is stuck on the log in message 
3. Log out and log in to a different account with no sites, notice both widgets showing incorrect info: the Views one shows data for the previously logged in site (although this seems to correct itself after a minute) and the Likes & Comments one shows a Log in message

These are all fixed by re-adding the widgets, and I'm not sure they're blockers for this PR but may be good to look into before shipping.

staskus commented:
@guarani thanks for testing!

> I noticed a few scenarios where the widgets got "stuck" showing an outdated message.

Adding it to the todo list to check!
